foreman/data_refinery_foreman/foreman/main.py

```
@@ -0,0 +1,260 @@
+    handle_downloader_jobs(failed_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_hung_downloader_jobs() -> None:
+    """Retry downloader jobs that were started but never finished."""
+    minimum_start_time = timezone.now() - MAX_DOWNLOADER_RUN_TIME
+    hung_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        end_time=None,
+        start_time__lt=minimum_start_time
+    )
+
+    handle_downloader_jobs(hung_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_lost_downloader_jobs() -> None:
+    """Retry downloader jobs that went too long without being started.
+
+    Idea: at some point this function could integrate with the spot
+    instances to determine if jobs are hanging due to a lack of
+    instances. A naive time-based implementation like this could end
+    up retrying every single queued job if there were a long period of
+    spot instance bid price being very high.
+    """
+    minimum_creation_time = timezone.now() - MAX_QUEUE_TIME
+    lost_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        start_time=None,
+        end_time=None,
+        created_at__lt=minimum_creation_time
+    )
+
+    handle_downloader_jobs(lost_jobs)
+
+
+@retry(stop_max_attempt_number=3)
+@transaction.atomic
+def requeue_processor_job(last_job: ProcessorJob) -> None:
+    """Queues a new downloader job and a Celery task for it.
+
+    The new downloader job will have num_retries one greater than
+    last_job.num_retries.
+    """
+    num_retries = last_job.num_retries + 1
+
+    new_job = ProcessorJob.create_job_and_relationships(num_retries=num_retries,
+                                                        batches=list(last_job.batches.all()),
+                                                        pipeline_applied=last_job.pipeline_applied)
+    processor_task = PROCESSOR_PIPELINE_LOOKUP[last_job.pipeline_applied]
```

The only way this error could happen is if a processor was added to the workers project without also being added here, which would be a mistake. If that were to happen I would want this to fail loudly and spectacularly so it immediately became obvious. I don't think this is an error that would come up in a normal scenario.